README.md
[![Build Status](https://travis-ci.org/benmeyer50/Finance-btce.png?branch=master)](https://travis-ci.org/benmeyer50/Finance-btce)

Finance-btce version 0.04
=========================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install


<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#Version">Version</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#METHODS">METHODS</a>
    <ul>
      <li><a href="#new">new</a></li>
      <li><a href="#getTicker">getTicker</a></li>
      <li><a href="#getInfo">getInfo</a></li>
      <li><a href="#activeOrders">activeOrders</a></li>
      <li><a href="#trade">trade</a></li>
      <li><a href="#cancelOrder">cancelOrder</a></li>
      <li><a href="#EXPORT">EXPORT</a></li>
    </ul>
  </li>
  <li><a href="#BUGS">BUGS</a></li>
  <li><a href="#AUTHOR">AUTHOR</a></li>
  <li><a href="#COPYRIGHT-AND-LICENSE">COPYRIGHT AND LICENSE</a></li>
</ul>

<h1 id="NAME">NAME</h1>

<p>Finance::btce - Perl extension for interfacing with the BTC-e bitcoin exchange</p>

<h1 id="Version">Version</h1>

<p>Version 0.04</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<pre><code>  use Finance::btce;

  my $btce = Finance::btce-&gt;new({key =&gt; &#39;key&#39;, secret =&gt; &#39;secret&#39;,});</code></pre>

<p>#public API calls</p>

<pre><code>  #Prices for Bitcoin to USD
  my %price = %{BTCtoUSD()};

  #Prices for Litecoin to Bitcoin
  my %price = %{LTCtoBTC()};
  
  #Prices for Litecoin to USD
  my %price = %{LTCtoUSD()};</code></pre>

<p>#with objects</p>

<pre><code>  #Prices for Bitcoin to USD
  $btce-&gt;getTicker(&#39;btc_usd&#39;);
  
  #account info
  $btce-&gt;getInfo();
  
  #make a trade
  $btce-&gt;trade(&#39;btc_usd&#39;, &#39;sell&#39;, 705.5, 1.22);</code></pre>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p><code>Finance::btce</code> is a wrapper module for the btc-e.com webservice.</p>

<h1 id="METHODS">METHODS</h1>

<h2 id="new">new</h2>

<p>Create a new <code>Finance::btce</code> object with your api <code>apikey</code> and <code>secret</code>.</p>

<pre><code>  my $btce = Finance::btce-&gt;new(apikey =&gt; &#39;xxx&#39;, secret =&gt; &#39;xxx&#39;);
  </code></pre>

<h2 id="getTicker">getTicker</h2>

<p>Gets current ticker data from API according to pair EX. btc_usd</p>

<pre><code>  $btce-&gt;getTicker(&#39;btc_usd&#39;);
  </code></pre>

<p>Returns:</p>

<pre><code>  $VAR1 = {
          &#39;vol_cur&#39; =&gt; &#39;16320.89588&#39;,
          &#39;vol&#39; =&gt; &#39;17038475.28297&#39;,
          &#39;avg&#39; =&gt; &#39;1047.301485&#39;,
          &#39;last&#39; =&gt; 1045,
          &#39;sell&#39; =&gt; 1045,
          &#39;buy&#39; =&gt; &#39;1045.039&#39;,
          &#39;high&#39; =&gt; &#39;1095.19995&#39;,
          &#39;server_time&#39; =&gt; 1385845491,
          &#39;low&#39; =&gt; &#39;999.40302&#39;,
          &#39;updated&#39; =&gt; 1385845490
  };
  </code></pre>

<p>On error, returns undef.</p>

<h2 id="getInfo">getInfo</h2>

<p>Gets current account info from API</p>

<pre><code>  $btce-&gt;getInfo();
  </code></pre>

<p>Returns:</p>

<pre><code>  $VAR1 = {
          &#39;success&#39; =&gt; 1,
          &#39;return&#39; =&gt; {
                        &#39;rights&#39; =&gt; {
                                      &#39;info&#39; =&gt; 1,
                                      &#39;withdraw&#39; =&gt; 0,
                                      &#39;trade&#39; =&gt; 0
                                    },
                        &#39;funds&#39; =&gt; {
                                     &#39;nvc&#39; =&gt; 0,
                                     &#39;nmc&#39; =&gt; 0,
                                     &#39;btc&#39; =&gt; &#39;0.53694401&#39;,
                                     &#39;xpm&#39; =&gt; 0,
                                     &#39;usd&#39; =&gt; &#39;0.35863074&#39;,
                                     &#39;ftc&#39; =&gt; 0,
                                     &#39;ltc&#39; =&gt; 0,
                                     &#39;trc&#39; =&gt; 0,
                                     &#39;rur&#39; =&gt; 0,
                                     &#39;ppc&#39; =&gt; 0,
                                     &#39;eur&#39; =&gt; 0
                                   },
                        &#39;server_time&#39; =&gt; 1385845686,
                        &#39;open_orders&#39; =&gt; 1,
                        &#39;transaction_count&#39; =&gt; 108
                      }
  };
  </code></pre>

<p>On error, returns undef.</p>

<h2 id="activeOrders">activeOrders</h2>

<p>Gets any active orders.</p>

<pre><code>  $btce-&gt;activeOrders();
  </code></pre>

<p>Returns a hash by order_id:</p>

<pre><code>  $VAR1 = {
          &#39;success&#39; =&gt; 1,
          &#39;return&#39; =&gt; {
                        &#39;72804751&#39; =&gt; {
                                        &#39;rate&#39; =&gt; &#39;30&#39;,
                                        &#39;timestamp_created&#39; =&gt; 1385776149,
                                        &#39;amount&#39; =&gt; &#39;73.966&#39;,
                                        &#39;pair&#39; =&gt; &#39;ltc_usd&#39;,
                                        &#39;status&#39; =&gt; 0,
                                        &#39;type&#39; =&gt; &#39;buy&#39;
                                      }
                      }
  };
  </code></pre>

<p>On error, returns undef</p>

<h2 id="trade">trade</h2>

<p>Places a trade on the market by pair (ex. btc_usd), type (ex. sell or buy), rate (ex. 980.1), and amount (ex. 2)</p>

<pre><code>  $btce-&gt;trade(&#39;btc_usd&#39;, &#39;buy&#39;, 980.1, 2);
  </code></pre>

<p>On success, returns the following hash:</p>

<pre><code>  $VAR1 = {
          &#39;success&#39; =&gt; 1,
          &#39;return&#39; =&gt; {
                        &#39;funds&#39; =&gt; {
                                     &#39;nvc&#39; =&gt; 0,
                                     &#39;nmc&#39; =&gt; 0,
                                     &#39;btc&#39; =&gt; &#39;1.01803021&#39;,
                                     &#39;xpm&#39; =&gt; 0,
                                     &#39;usd&#39; =&gt; &#39;0.30042416&#39;,
                                     &#39;ftc&#39; =&gt; 0,
                                     &#39;ltc&#39; =&gt; &#39;6.98292&#39;,
                                     &#39;trc&#39; =&gt; 0,
                                     &#39;rur&#39; =&gt; 0,
                                     &#39;ppc&#39; =&gt; 0,
                                     &#39;eur&#39; =&gt; 0
                                   },
                        &#39;remains&#39; =&gt; &#39;1.1&#39;,
                        &#39;order_id&#39; =&gt; 79247481,
                        &#39;received&#39; =&gt; 0
                      }
  };
  </code></pre>

<p>On error, returns undef</p>

<h2 id="cancelOrder">cancelOrder</h2>

<p>Cancel an active order.</p>

<pre><code>  $btce-&gt;cancelOrder(&#39;79214411&#39;);
  </code></pre>

<p>Returns a hash with order_id and funds.</p>

<pre><code>  $VAR1 = {
          &#39;success&#39; =&gt; 1,
          &#39;return&#39; =&gt; {
                        &#39;funds&#39; =&gt; {
                                     &#39;nvc&#39; =&gt; 0,
                                     &#39;nmc&#39; =&gt; 0,
                                     &#39;btc&#39; =&gt; &#39;1.01803021&#39;,
                                     &#39;xpm&#39; =&gt; 0,
                                     &#39;usd&#39; =&gt; &#39;0.30042416&#39;,
                                     &#39;ftc&#39; =&gt; 0,
                                     &#39;ltc&#39; =&gt; &#39;8.08292&#39;,
                                     &#39;trc&#39; =&gt; 0,
                                     &#39;rur&#39; =&gt; 0,
                                     &#39;ppc&#39; =&gt; 0,
                                     &#39;eur&#39; =&gt; 0
                                   },
                        &#39;order_id&#39; =&gt; 79214411
                      }
  };
  </code></pre>

<p>On error, returns undef</p>

<h2 id="EXPORT">EXPORT</h2>

<p>None by default.</p>

<h1 id="BUGS">BUGS</h1>

<p>Please report all bug and feature requests through github at <a href="https://github.com/benmeyer50/Finance-btce/issues">https://github.com/benmeyer50/Finance-btce/issues</a></p>

<h1 id="AUTHOR">AUTHOR</h1>

<p>Benjamin Meyer, &lt;bmeyer@benjamindmeyer.com&gt;</p>

<h1 id="COPYRIGHT-AND-LICENSE">COPYRIGHT AND LICENSE</h1>

<p>Copyright (C) 2013 by Benjamin Meyer</p>

<p>This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself, either Perl version 5.12.4 or, at your option, any later version of Perl 5 you may have available.</p>





Changes
Revision history for Perl extension Finance::btce.

0.04 Sun Jun 29 00:07:29 2014
	- Use HTTP::Tiny, remove WWW::Mechanize

0.03 Wed Jan 1 17:45:07 2014
	- complete rewrite.

0.02  Tue Apr 23 03:50:00 2013
	- added all public api tickers
	- added new() method to store keys
	- added getInfo method to return info about account

0.01  Mon Apr 22 06:44:48 2013
	- original version; created by h2xs 1.23 with options
		-AX -n Finance::btce




lib/Finance/btce.pm
package Finance::btce;

use 5.012004;
use JSON;
use Carp qw(croak carp);
use Digest::SHA qw( hmac_sha512_hex);
use HTTP::Tiny;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Finance::btce ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(BTCtoUSD LTCtoBTC LTCtoUSD ) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw();

our $VERSION = '0.04';

our $post_url = "https://btc-e.com/tapi";
our $ws_url = "https://btc-e.com/api/2/";

sub BTCtoUSD
{
	my $browser = HTTP::Tiny->new();
	my $json = JSON->new->allow_nonref;
	my $ticker;
	my $res;
	eval { $res = $browser->get("$ws_url" . "btc_usd/ticker"); }; carp $@ if $@;
	if(not $res->{success}){
		carp "No Response received.\n";
		return undef;
	}

	eval { $ticker = $json->decode($res->{content})->{ticker}; }; carp $@ if $@;
	return $ticker;
}

sub LTCtoBTC
{
	my $browser = HTTP::Tiny->new();
	my $json = JSON->new->allow_nonref;
	my $ticker;
	my $res;
	eval { $res = $browser->get("$ws_url" . "ltc_btc/ticker"); }; carp $@ if $@;
	if(not $res->{success}){
		carp "No Response received.\n";
		return undef;
	}

	eval { $ticker = $json->decode($res->{content})->{ticker}; }; carp $@ if $@;
	return $ticker;
}

sub LTCtoUSD
{
	my $browser = HTTP::Tiny->new();
	my $json = JSON->new->allow_nonref;
	my $ticker;
	my $res;
	eval { $res = $browser->get("$ws_url" . "ltc_usd/ticker"); }; carp $@ if $@;
	if(not $res->{success}){
		carp "No Response received.\n";
		return undef;
	}

	eval { $ticker = $json->decode($res->{content})->{ticker}; }; carp $@ if $@;
	return $ticker;
}


### Authenticated API calls

sub new
{
	my ($class, $args) = @_;
	if($args->{'apikey'} && $args->{'secret'})
	{
		$args->{mech} = HTTP::Tiny->new( timeout => 4);
		$args->{json} = JSON->new();
		$args->{nonce} = _nonce_closure();
	}
	else
	{
		croak "You must provide an apikey and secret";
	}
	return bless $args, $class;
}


sub getTicker
{
	my ($self, $pair) = @_;
	my $ticker;
	my $res;
	unless($pair){
		carp "Must pass valid currency pair. Ex: btc_usd\n";
		return undef;
	}
	my $url = $ws_url . "$pair/" . "ticker";
	eval { $res = $self->_mech->get($url);}; carp $@ if $@;
	if(not $res->{success} ){
		carp "No success getting data from  $url $res->{reason} $res->{content}\n";
		return undef;
	}
	eval { $ticker = $self->_json->decode($res->{content})->{ticker}; }; carp $@ if $@;
	return $ticker;
}


sub getInfo
{
	my ($self) = @_;
	my $nonce = $self->_get_nonce();
	my $data = "method=getInfo&nonce=$nonce";
	my $res;
	my $headers = {headers => {'Key' => $self->_apikey, 'Sign' => $self->_hash($data)} };

	eval {$res = $self->_mech->post_form($post_url, ['method' => 'getInfo', 'nonce' => $nonce], $headers); }; carp $@ if $@;
	if(not $res->{success} ){
		carp "No success posting date to $post_url: $res->{reason} $res->{content}\n";
		return undef;
	}
	my $ref =  $self->_json->decode($res->{content});

	if($ref->{success} != 1){
		carp "Api did not return 'success : 1' from post to $post_url\n";
		return undef;
	}
	return $ref;
}


sub activeOrders
{
  	my ($self) = @_;
	my $nonce = $self->_get_nonce();
	my $data = "method=ActiveOrders&nonce=$nonce";
	my $res;
	my $headers = {headers => {'Key' => $self->_apikey, 'Sign' => $self->_hash($data)} };

	eval {$res = $self->_mech->post_form($post_url, ['method' => 'ActiveOrders', 'nonce' => $nonce], $headers); }; carp $@ if $@;
	if(not $res->{success} ){
		carp "No success posting data to $post_url\n";
		return undef;
	}
	my $ref =  $self->_json->decode($res->{content});
	if($ref->{success} != 1 and $ref->{error} ne "no orders"){
		carp "Api did not return 'success : 1' from post to $post_url\n" .
			"Error was: $ref->{error}\n";
		return undef;
	}
	return $ref;

}


sub trade
{
	my ($self, $pair, $type, $rate, $amount) = @_;
	my $nonce = $self->_get_nonce();
	my $data = "method=Trade&nonce=$nonce&pair=$pair&type=$type&rate=$rate&amount=$amount";
	my $res;
	unless($pair and $type and $rate and $amount){
		carp "Must pass valid currency pair (ex.btc_usd), type (buy or sell), rate and amount.\n";
		return undef;
	}
	my $headers = {headers => {'Key' => $self->_apikey, 'Sign' => $self->_hash($data)} };
	
	eval { $res = $self->_mech->post_form($post_url, ['method' => 'Trade', 'nonce' => $nonce, 'pair' => $pair, 
		'type' => $type, 'rate' => $rate, 'amount' => $amount], $headers); }; carp $@ if $@;
	if(not $res->{success} ){
		carp "No success posting data to $post_url\n";
		return undef;
	}
	my $ref =  $self->_json->decode($res->{content});

	if($ref->{success} != 1){
		carp "Api did not return 'success : 1' from post to $post_url\n" .
			"Error was: $ref->{error}\n";
		return undef;
	}

	return $ref;
}


sub cancelOrder
{
	my ($self,$order_id) = @_;
	my $nonce = $self->_get_nonce();
	my $data = "method=CancelOrder&nonce=$nonce&order_id=$order_id";
	my $res;
	my $headers = {headers => {'Key' => $self->_apikey, 'Sign' => $self->_hash($data)} };

	eval { $res = $self->_mech->post_form($post_url, ['method' => 'CancelOrder', 'nonce' => $nonce, 'order_id' => $order_id], $headers); }; carp $@ if $@;
	if(not $res->{success} ){
		carp "No success posting data to $post_url\n";
		return undef;
	}
	my $ref =  $self->_json->decode($res->{content});
	if($ref->{success} != 1){
		carp "Api did not return 'success : 1' from post to $post_url\n" .
			"Error was: $ref->{error}\n";
		return undef;
	}

	return $ref;
}


#private methods

sub _apikey
{
	my ($self) = @_;
	return $self->{'apikey'};
}

sub _secretkey
{
	my ($self) = @_;
	return $self->{'secret'};
}

sub _mech
{
	my $self = shift();
	return $self->{mech};
}

sub _json
{
	my $self = shift();
	return $self->{json};
}

sub _hash
{
	my ($self,$data) = @_;
	return hmac_sha512_hex($data, $self->_secretkey() );
}

sub _nonce_closure
{
	my $nonce = time();
	return sub { $nonce++ };
}

sub _get_nonce
{
	my $self = shift();
	return $self->{nonce}->();
}

1;

__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

Finance::btce - Perl extension for interfacing with the BTC-e bitcoin exchange

=head1 Version

Version 0.021

=head1 SYNOPSIS

  use Finance::btce;

  my $btce = Finance::btce->new({key => 'key', secret => 'secret',});

#public API calls
  
  #Prices for Bitcoin to USD
  my %price = %{BTCtoUSD()};

  #Prices for Litecoin to Bitcoin
  my %price = %{LTCtoBTC()};
  
  #Prices for Litecoin to USD
  my %price = %{LTCtoUSD()};

#with objects

  #Prices for Bitcoin to USD
  $btce->getTicker('btc_usd');
  
  #account info
  $btce->getInfo();
  
  #make a trade
  $btce->trade('btc_usd', 'sell', 705.5, 1.22);


=head1 DESCRIPTION

C<Finance::btce> is a wrapper module for the btc-e.com webservice.

=head1 METHODS

=head2 new

Create a new C<Finance::btce> object with your api C<apikey> and C<secret>.

  
  my $btce = Finance::btce->new(apikey => 'xxx', secret => 'xxx');
  

=head2 getTicker

Gets current ticker data from API according to pair EX. btc_usd

  
  $btce->getTicker('btc_usd');
  

Returns:

  $VAR1 = {
          'vol_cur' => '16320.89588',
          'vol' => '17038475.28297',
          'avg' => '1047.301485',
          'last' => 1045,
          'sell' => 1045,
          'buy' => '1045.039',
          'high' => '1095.19995',
          'server_time' => 1385845491,
          'low' => '999.40302',
          'updated' => 1385845490
  };
  

On error, returns undef.


=head2 getInfo

Gets current account info from API

  
  $btce->getInfo();
  

Returns:

  $VAR1 = {
          'success' => 1,
          'return' => {
                        'rights' => {
                                      'info' => 1,
                                      'withdraw' => 0,
                                      'trade' => 0
                                    },
                        'funds' => {
                                     'nvc' => 0,
                                     'nmc' => 0,
                                     'btc' => '0.53694401',
                                     'xpm' => 0,
                                     'usd' => '0.35863074',
                                     'ftc' => 0,
                                     'ltc' => 0,
                                     'trc' => 0,
                                     'rur' => 0,
                                     'ppc' => 0,
                                     'eur' => 0
                                   },
                        'server_time' => 1385845686,
                        'open_orders' => 1,
                        'transaction_count' => 108
                      }
  };
  
On error, returns undef.

=head2 activeOrders

Gets any active orders.

  
  $btce->activeOrders();
  

Returns a hash by order_id:

  $VAR1 = {
          'success' => 1,
          'return' => {
                        '72804751' => {
                                        'rate' => '30',
                                        'timestamp_created' => 1385776149,
                                        'amount' => '73.966',
                                        'pair' => 'ltc_usd',
                                        'status' => 0,
                                        'type' => 'buy'
                                      }
                      }
  };
  
On error, returns undef

=head2 trade

Places a trade on the market by pair (ex. btc_usd), type (ex. sell or buy), rate (ex. 980.1), and amount (ex. 2)

  
  $btce->trade('btc_usd', 'buy', 980.1, 2);
  

On success, returns the following hash:

  $VAR1 = {
          'success' => 1,
          'return' => {
                        'funds' => {
                                     'nvc' => 0,
                                     'nmc' => 0,
                                     'btc' => '1.01803021',
                                     'xpm' => 0,
                                     'usd' => '0.30042416',
                                     'ftc' => 0,
                                     'ltc' => '6.98292',
                                     'trc' => 0,
                                     'rur' => 0,
                                     'ppc' => 0,
                                     'eur' => 0
                                   },
                        'remains' => '1.1',
                        'order_id' => 79247481,
                        'received' => 0
                      }
  };
  

On error, returns undef

=head2 cancelOrder

Cancel an active order.

  
  $btce->cancelOrder('79214411');
  

Returns a hash with order_id and funds.

  $VAR1 = {
          'success' => 1,
          'return' => {
                        'funds' => {
                                     'nvc' => 0,
                                     'nmc' => 0,
                                     'btc' => '1.01803021',
                                     'xpm' => 0,
                                     'usd' => '0.30042416',
                                     'ftc' => 0,
                                     'ltc' => '8.08292',
                                     'trc' => 0,
                                     'rur' => 0,
                                     'ppc' => 0,
                                     'eur' => 0
                                   },
                        'order_id' => 79214411
                      }
  };
  
On error, returns undef

=head2 EXPORT

None by default.

=head1 BUGS

Please report all bug and feature requests through github
at L<https://github.com/benmeyer50/Finance-btce/issues>

=head1 AUTHOR

Benjamin Meyer, E<lt>bmeyer@benjamindmeyer.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2013 by Benjamin Meyer

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.12.4 or,
at your option, any later version of Perl 5 you may have available.


=cut
